File: src/Javascript/array/34.在排序数组中查找元素的第一个和最后一个位置.js
/*
 * @lc app=leetcode.cn id=34 lang=javascript
 *
 * [34] 在排序数组中查找元素的第一个和最后一个位置
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function(nums, target) {
    // find the first element bigger or equal to target
    const leftIndex = binarySearch(nums, target, true);
    // find the first element bigger than target
    const rightIndex = binarySearch(nums, target, false) - 1;
    // boader protection
    if (leftIndex >= 0 && leftIndex <= rightIndex && rightIndex < nums.length && nums[leftIndex] === target && nums[rightIndex] === target)
    return [leftIndex, rightIndex];
    return [-1, -1];
};

// normal binary search
const binarySearch = (nums, target, lower) => {
    let left = 0, right = nums.length - 1, ans = nums.length;
    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        if (nums[mid] > target || (lower && nums[mid] >= target)) {
            right = mid - 1;
            ans = mid;
        } else left = mid + 1;
    }
    return ans;
}
// @lc code=end

File: src/Java/binary-search/34.在排序数组中查找元素的第一个和最后一个位置.java
/*
 * @lc app=leetcode.cn id=34 lang=java
 *
 * [34] 在排序数组中查找元素的第一个和最后一个位置
 */

// @lc code=start
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int leftIndex = binarySearch(nums, target, true);
        int rightIndex = binarySearch(nums, target, false) - 1;
        if (leftIndex >= 0 && leftIndex <= rightIndex && rightIndex < nums.length && nums[leftIndex]  == target && nums[rightIndex] == target)
            return new int[]{leftIndex, rightIndex};
        return new int[]{-1, -1};
    }
    public int binarySearch(int[] nums, int target, boolean flag) {
        int left = 0, right = nums.length - 1, ans = nums.length;
        while (left <= right) {
            int mid = (left + right) / 2;
            if (nums[mid] > target || flag && nums[mid] >= target) {
                ans = mid;
                right = mid - 1;
            } else left = mid + 1;
        }
        return ans;
    }
}
// @lc code=end

